erdma controller 当前只支持主网卡eth0和辅助网卡是erdma，不支持两个erdma都是辅助网卡

erdma controller 当前只支持主网卡eth0和辅助网卡是erdma，不支持两个erdma都是辅助网卡。会报如下错误：

panic: probe device failed, err: get erdma link failed: cannot found rdma link for eth0

goroutine 1 [running]:
main.main()
	/workspace/cmd/agent/main.go:36 +0x22f


# ibv_devinfo
hca_id:	erdma_0
	transport:			eRDMA (0)
	fw_ver:				0.2.0
	node_guid:			0216:3eff:fe18:e894
	sys_image_guid:			0216:3eff:fe18:e894
	vendor_id:			0x1ded
	vendor_part_id:			4223
	hw_ver:				0x0
	phys_port_cnt:			1
		port:	1
			state:			PORT_ACTIVE (4)
			max_mtu:		1024 (3)
			active_mtu:		1024 (3)
			sm_lid:			0
			port_lid:		0
			port_lmc:		0x00
			link_layer:		Ethernet

hca_id:	erdma_1
	transport:			eRDMA (0)
	fw_ver:				0.2.0
	node_guid:			0216:3eff:fe18:db25
	sys_image_guid:			0216:3eff:fe18:db25
	vendor_id:			0x1ded
	vendor_part_id:			4223
	hw_ver:				0x0
	phys_port_cnt:			1
		port:	1
			state:			PORT_ACTIVE (4)
			max_mtu:		1024 (3)
			active_mtu:		1024 (3)
			sm_lid:			0
			port_lid:		0
			port_lmc:		0x00
			link_layer:		Ethernet

# ibv_devices
    device          	   node GUID
    ------          	----------------
    erdma_0         	02163efffe18e894
    erdma_1         	02163efffe18db25
# ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.53.4.141  netmask 255.255.255.0  broadcast 10.53.4.255
        inet6 fe80::216:3eff:fe04:8159  prefixlen 64  scopeid 0x20<link>
        ether 00:16:3e:04:81:59  txqueuelen 1000  (Ethernet)
        RX packets 4908192200  bytes 5593465005686 (5.5 TB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1247414571  bytes 43365575462054 (43.3 TB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.53.5.201  netmask 255.255.255.0  broadcast 10.53.5.255
        inet6 fe80::216:3eff:fe18:e894  prefixlen 64  scopeid 0x20<link>
        ether 00:16:3e:18:e8:94  txqueuelen 1000  (Ethernet)
        RX packets 134694  bytes 5659340 (5.6 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 134908  bytes 5673584 (5.6 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth2: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.53.5.200  netmask 255.255.255.0  broadcast 10.53.5.255
        inet6 fe80::216:3eff:fe18:db25  prefixlen 64  scopeid 0x20<link>
        ether 00:16:3e:18:db:25  txqueuelen 1000  (Ethernet)
        RX packets 140734  bytes 8541874 (8.5 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 150752  bytes 29865937 (29.8 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

--- a/go.mod
+++ b/go.mod
@@ -9,6 +9,7 @@ require (
 	github.com/onsi/ginkgo/v2 v2.19.0
 	github.com/onsi/gomega v1.33.1
 	github.com/samber/lo v1.47.0
+	github.com/stretchr/testify v1.9.0
 	github.com/vishvananda/netlink v1.3.0
 	golang.org/x/net v0.26.0
 	google.golang.org/grpc v1.65.0
@@ -23,6 +24,7 @@ require (
 
 require (
 	github.com/morikuni/aec v1.0.0 // indirect
+	github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2 // indirect
 	gotest.tools/v3 v3.5.1 // indirect
 )
 

--- a/internal/controller/eri.go
+++ b/internal/controller/eri.go
@@ -2,6 +2,7 @@ package controller
 
 import (
 	"fmt"
+	"os"
 	"strings"
 
 	networkv1 "github.com/AliyunContainerService/alibabacloud-erdma-controller/api/v1"
@@ -32,7 +33,7 @@ const (
 type EriClient struct {
 	client          *ecs.Client
 	regionID        string
-	managedNonOwned bool
+	ManagedNonOwned bool
 }
 
 func NewEriClient(k8sClient client.Client) (*EriClient, error) {
@@ -40,8 +41,12 @@ func NewEriClient(k8sClient client.Client) (*EriClient, error) {
 	if err != nil {
 		return nil, err
 	}
+	network := "tcp"
+	if os.Getenv("PUBLIC_NETWORK") == "true" {
+		network = "public"
+	}
 
-	ecsEndpoint, err := service.GetEndpointRules(tea.String("ecs"), tea.String(config.GetConfig().Region), tea.String("regional"), tea.String("vpc"), nil)
+	ecsEndpoint, err := service.GetEndpointRules(tea.String("ecs"), tea.String(config.GetConfig().Region), tea.String("regional"), tea.String(network), nil)
 	if err != nil {
 		return nil, err
 	}
@@ -50,7 +55,7 @@ func NewEriClient(k8sClient client.Client) (*EriClient, error) {
 		UserAgent:    ptr.To("AlibabaCloud/ERdma-Controller/0.1"),
 		Credential:   cred,
 		EndpointType: tea.String("regional"),
-		Network:      tea.String("vpc"),
+		Network:      tea.String(network),
 		Endpoint:     ecsEndpoint,
 	})
 	if err != nil {
@@ -58,7 +63,7 @@ func NewEriClient(k8sClient client.Client) (*EriClient, error) {
 	}
 	return &EriClient{
 		regionID:        config.GetConfig().Region,
-		managedNonOwned: config.GetConfig().ManageNonOwnedERIs,
+		ManagedNonOwned: config.GetConfig().ManageNonOwnedERIs,
 		client:          client,
 	}, nil
 }
@@ -173,7 +178,7 @@ func (e *EriClient) ConvertPrimaryENI(primaryENI string, queuePair int) error {
 		NetworkInterfaceTrafficConfig: &ecs.ModifyNetworkInterfaceAttributeRequestNetworkInterfaceTrafficConfig{
 			NetworkInterfaceTrafficMode: ptr.To(trafficModeRDMA),
 			// todo: not support dynamic set queue pair number
-			// QueuePairNumber:             ptr.To(int32(queuePair)),
+			QueuePairNumber: ptr.To(int32(queuePair)),
 		},
 	}); err != nil {
 		return err
@@ -218,9 +223,13 @@ func (e *EriClient) SelectERIs(instanceID string) ([]*types.ERI, error) {
 			cardCount = int(min(*instanceType.NetworkCardQuantity, eriQuantity))
 		}
 		queuePairCount = int(*instanceType.QueuePairNumber)
+		// GPU instance max queue pair number is card count * queue pair number
+		if instanceType.GPUAmount != nil && *instanceType.GPUAmount > 0 {
+			queuePairCount = int(*instanceType.QueuePairNumber * int32(cardCount))
+		}
 	}
 
-	existENIs, err := e.client.DescribeNetworkInterfaces(&ecs.DescribeNetworkInterfacesRequest{
+	describeENIResponse, err := e.client.DescribeNetworkInterfaces(&ecs.DescribeNetworkInterfacesRequest{
 		RegionId:   ptr.To(e.regionID),
 		InstanceId: ptr.To(instanceID),
 		PageSize:   ptr.To(int32(100)),
@@ -228,45 +237,88 @@ func (e *EriClient) SelectERIs(instanceID string) ([]*types.ERI, error) {
 	if err != nil {
 		return nil, fmt.Errorf("cannot found node eni: %v", err)
 	}
+	existENIs := describeENIResponse.Body.NetworkInterfaceSets.NetworkInterfaceSet
+	selectEriList, needCreate, queuePairNumberConfig, err := e.SelectEriFromExist(existENIs, queuePairCount, cardCount)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate eri config list from exist enis: %v", err)
+	}
+	eris, err := e.CreateEriForInstance(instanceResp.Body.Instances.Instance[0], needCreate, queuePairNumberConfig)
+	if err != nil {
+		return nil, err
+	}
+	selectEriList = append(selectEriList, eris...)
+
+	return selectEriList, nil
+}
+
+func (e *EriClient) SelectEriFromExist(existENIs []*ecs.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet, queuePairCount, cardCount int) ([]*types.ERI, []int, int, error) {
+	var existQueuePairCount int
+	existERIs := lo.Filter(existENIs, func(item *ecs.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet, _ int) bool {
+		eri := item.NetworkInterfaceTrafficMode != nil && *item.NetworkInterfaceTrafficMode == trafficModeRDMA
+		if eri {
+			existQueuePairCount += int(*item.QueuePairNumber)
+		}
+		return eri
+	})
+	eriLog.Info("exist eri", "existERIs", lo.Map(existERIs, func(item *ecs.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet, _ int) *types.ERI {
+		return toEri(item, 0)
+	}), "existQueuePairCount", existQueuePairCount, "osMaxQueuePairCount", queuePairCount, "cardCount", cardCount)
+
 	var (
 		selectedENIs []*ecs.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet
 		cardIndexENI = map[int]*types.ERI{}
 	)
-	for _, eni := range existENIs.Body.NetworkInterfaceSets.NetworkInterfaceSet {
-		if eni.Type != nil && *eni.Type == "Primary" {
-			selectedENIs = append(selectedENIs, eni)
-			cardIndexENI[0] = toEri(eni, queuePairCount/cardCount)
-		} else {
-			if eni.NetworkInterfaceTrafficMode != nil && *eni.NetworkInterfaceTrafficMode == trafficModeRDMA && e.OwnENI(eni) {
-				eniIndex := eniCardIndex(eni)
-				if _, ok := cardIndexENI[eniIndex]; !ok {
-					cardIndexENI[eniIndex] = toEri(eni, queuePairCount/cardCount)
-					selectedENIs = append(selectedENIs, eni)
-				}
+
+	for _, eri := range existERIs {
+		if !e.OwnENI(eri) {
+			continue
+		}
+		eniIndex := eniCardIndex(eri)
+		if _, ok := cardIndexENI[eniIndex]; !ok {
+			cardIndexENI[eniIndex] = toEri(eri, 0)
+			selectedENIs = append(selectedENIs, eri)
+		}
+	}
+	var needCreateOrConvert []int
+	if existQueuePairCount <= queuePairCount {
+		for i := 0; i < cardCount; i++ {
+			if _, ok := cardIndexENI[i]; !ok {
+				needCreateOrConvert = append(needCreateOrConvert, i)
 			}
 		}
 	}
-	if len(cardIndexENI) == 0 {
-		return nil, fmt.Errorf("cannot found node primary eni")
+
+	var remainQueuePairCountPerCardIndex int
+	if len(needCreateOrConvert) > 0 {
+		remainQueuePairCountPerCardIndex = (queuePairCount - existQueuePairCount) / len(needCreateOrConvert)
+		if remainQueuePairCountPerCardIndex > 0 {
+			if _, ok := cardIndexENI[0]; !ok {
+				// if cardIndex 0 not bind ENI, using primary ENI as cardIndex 0 ENI
+				for _, eni := range existENIs {
+					if eni.Type != nil && *eni.Type == "Primary" {
+						selectedENIs = append(selectedENIs, eni)
+						cardIndexENI[0] = toEri(eni, remainQueuePairCountPerCardIndex)
+						cardIndex0Idx := lo.IndexOf(needCreateOrConvert, 0)
+						// remove from create list
+						needCreateOrConvert = append(needCreateOrConvert[:cardIndex0Idx], needCreateOrConvert[cardIndex0Idx+1:]...)
+					}
+				}
+			}
+			if len(cardIndexENI) == 0 {
+				return nil, nil, 0, fmt.Errorf("cannot find node primary ENI or existing ENI")
+			}
+		} else {
+			needCreateOrConvert = nil
+		}
 	}
 
 	eriList := lo.Map(selectedENIs, func(item *ecs.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet, _ int) *types.ERI {
-		return toEri(item, queuePairCount/cardCount)
+		return toEri(item, remainQueuePairCountPerCardIndex)
 	})
-
-	var needCreate []int
-	for i := 0; i < cardCount; i++ {
-		if _, ok := cardIndexENI[i]; !ok {
-			needCreate = append(needCreate, i)
-		}
+	if len(eriList) == 0 && len(needCreateOrConvert) == 0 {
+		return nil, nil, 0, fmt.Errorf("cannot create ERI for instance due to no available slot")
 	}
-	eris, err := e.CreateEriForInstance(instanceResp.Body.Instances.Instance[0], needCreate, queuePairCount/cardCount)
-	if err != nil {
-		return nil, err
-	}
-	eriList = append(eriList, eris...)
-
-	return eriList, nil
+	return eriList, needCreateOrConvert, remainQueuePairCountPerCardIndex, nil
 }
 
 func (e *EriClient) EnsureEriForInstance(devices []networkv1.DeviceInfo) ([]networkv1.DeviceStatus, error) {
@@ -346,7 +398,7 @@ func (e *EriClient) EnsureEriForInstance(devices []networkv1.DeviceInfo) ([]netw
 }
 
 func (e *EriClient) OwnENI(eni *ecs.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet) bool {
-	if e.managedNonOwned {
+	if e.ManagedNonOwned {
 		return true
 	}
 	if eni.Tags == nil || eni.Tags.Tag == nil {

--- a/internal/controller/eri_test.go
+++ b/internal/controller/eri_test.go
@@ -1,0 +1,603 @@
+// Copyright 2023 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controller_test
+
+import (
+	"testing"
+
+	"github.com/samber/lo"
+
+	"github.com/AliyunContainerService/alibabacloud-erdma-controller/internal/controller"
+	"github.com/AliyunContainerService/alibabacloud-erdma-controller/internal/types"
+	"github.com/alibabacloud-go/ecs-20140526/v4/client"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSelectEriFromExist(t *testing.T) {
+	tests := []struct {
+		name                     string
+		existENIs                []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet
+		queuePairCount           int
+		cardCount                int
+		expectedERIs             []*types.ERI
+		expectedNeedCreate       []int
+		expectedQueuePairPerCard int
+		expectedError            bool
+	}{
+		{
+			name:                     "No existing RDMA ENIs and Primary ENI",
+			existENIs:                []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{},
+			queuePairCount:           8,
+			cardCount:                2,
+			expectedERIs:             []*types.ERI{},
+			expectedNeedCreate:       []int{0, 1},
+			expectedQueuePairPerCard: 4,
+			expectedError:            true,
+		},
+		{
+			name: "Existing Primary ENIs is RDMA and with sufficient queue pairs",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(8)),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      1,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-1",
+					CardIndex:    0,
+					QueuePair:    8,
+					IsPrimaryENI: true,
+					MAC:          "00:16:3e:00:00:00",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            false,
+		},
+		{
+			name: "Existing Secondary RDMA ENI is RDMA and with sufficient queue pairs",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-2"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(8)),
+					Type:                        lo.ToPtr("Secondary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      1,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-2",
+					CardIndex:    0,
+					QueuePair:    8,
+					IsPrimaryENI: false,
+					MAC:          "00:16:3e:00:00:01",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            false,
+		},
+		{
+			name: "Existing Secondary RDMA ENIs with sufficient queue pairs on cardIndex 1",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-2"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(8)),
+					Type:                        lo.ToPtr("Secondary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(1)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      2,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-2",
+					CardIndex:    1,
+					QueuePair:    8,
+					IsPrimaryENI: false,
+					MAC:          "00:16:3e:00:00:01",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            false,
+		},
+		{
+			name: "Primary ENI is RDMA and used as card index 0, need create new on card index 1",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-primary"),
+					Type:                        lo.ToPtr("Primary"),
+					QueuePairNumber:             lo.ToPtr(int32(6)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      2,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-primary",
+					IsPrimaryENI: true,
+					CardIndex:    0,
+					QueuePair:    6,
+					MAC:          "00:16:3e:00:00:00",
+				},
+			},
+			expectedNeedCreate:       []int{1},
+			expectedQueuePairPerCard: 2,
+			expectedError:            false,
+		},
+		{
+			name: "Secondary ENI is RDMA and on both cardIndex 0 and 1, no need to create or convert new",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-primary"),
+					Type:                        lo.ToPtr("Primary"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:03"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-secondary1"),
+					Type:                        lo.ToPtr("Secondary"),
+					QueuePairNumber:             lo.ToPtr(int32(6)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(1)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-secondary2"),
+					Type:                        lo.ToPtr("Secondary"),
+					QueuePairNumber:             lo.ToPtr(int32(2)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      2,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-secondary1",
+					IsPrimaryENI: false,
+					CardIndex:    1,
+					QueuePair:    6,
+					MAC:          "00:16:3e:00:00:00",
+				},
+				{
+					ID:           "eni-secondary2",
+					IsPrimaryENI: false,
+					CardIndex:    0,
+					QueuePair:    2,
+					MAC:          "00:16:3e:00:00:01",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            false,
+		},
+		{
+			name: "Secondary ENI is RDMA and both on cardIndex 0, no need to create or convert new",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-primary"),
+					Type:                        lo.ToPtr("Primary"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:03"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-secondary1"),
+					Type:                        lo.ToPtr("Secondary"),
+					QueuePairNumber:             lo.ToPtr(int32(6)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-secondary2"),
+					Type:                        lo.ToPtr("Secondary"),
+					QueuePairNumber:             lo.ToPtr(int32(2)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      2,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-secondary1",
+					IsPrimaryENI: false,
+					CardIndex:    0,
+					QueuePair:    6,
+					MAC:          "00:16:3e:00:00:00",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            false,
+		},
+		{
+			name: "Secondary ENI is RDMA and both on cardIndex 0 and cardIndex 1, but queue pair can not be divided",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-primary"),
+					Type:                        lo.ToPtr("Primary"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:03"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-secondary1"),
+					Type:                        lo.ToPtr("Secondary"),
+					QueuePairNumber:             lo.ToPtr(int32(6)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(1)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-secondary2"),
+					Type:                        lo.ToPtr("Secondary"),
+					QueuePairNumber:             lo.ToPtr(int32(2)),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(1)),
+					},
+				},
+			},
+			queuePairCount: 9,
+			cardCount:      3,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-secondary1",
+					IsPrimaryENI: false,
+					CardIndex:    1,
+					QueuePair:    6,
+					MAC:          "00:16:3e:00:00:00",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            false,
+		},
+		{
+			name: "Need create on cardIndex 1",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(2)),
+					Type:                        lo.ToPtr("Secondary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      2,
+			expectedERIs: []*types.ERI{
+				{
+					ID:        "eni-1",
+					CardIndex: 0,
+					MAC:       "00:16:3e:00:00:00",
+					QueuePair: 2,
+				},
+			},
+			expectedNeedCreate:       []int{1},
+			expectedQueuePairPerCard: 6,
+			expectedError:            false,
+		},
+		{
+			name: "Need convert on cardIndex 0",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(2)),
+					Type:                        lo.ToPtr("Secondary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(1)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-2"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      2,
+			expectedERIs: []*types.ERI{
+				{
+					ID:        "eni-1",
+					CardIndex: 1,
+					MAC:       "00:16:3e:00:00:00",
+					QueuePair: 2,
+				},
+				{
+					ID:           "eni-2",
+					CardIndex:    0,
+					IsPrimaryENI: true,
+					MAC:          "00:16:3e:00:00:01",
+					QueuePair:    6,
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 6,
+			expectedError:            false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			e := &controller.EriClient{
+				ManagedNonOwned: true,
+			}
+			eris, needCreate, queuePairPerCard, err := e.SelectEriFromExist(tt.existENIs, tt.queuePairCount, tt.cardCount)
+			if tt.expectedError {
+				assert.Error(t, err)
+			} else {
+				if needCreate == nil {
+					needCreate = []int{}
+				}
+				assert.NoError(t, err)
+				assert.Equal(t, tt.expectedERIs, eris)
+				assert.Equal(t, tt.expectedNeedCreate, needCreate)
+				assert.Equal(t, tt.expectedQueuePairPerCard, queuePairPerCard)
+			}
+		})
+	}
+}
+
+func TestSelectEriNoManagedExists(t *testing.T) {
+	tests := []struct {
+		name                     string
+		existENIs                []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet
+		queuePairCount           int
+		cardCount                int
+		expectedERIs             []*types.ERI
+		expectedNeedCreate       []int
+		expectedQueuePairPerCard int
+		expectedError            bool
+	}{
+		{
+			name:                     "No existing RDMA ENIs and Primary ENI",
+			existENIs:                []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{},
+			queuePairCount:           8,
+			cardCount:                2,
+			expectedERIs:             []*types.ERI{},
+			expectedNeedCreate:       []int{0, 1},
+			expectedQueuePairPerCard: 4,
+			expectedError:            true,
+		},
+		{
+			name: "Existing Primary ENIs is RDMA and with sufficient queue pairs",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(8)),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount:           8,
+			cardCount:                1,
+			expectedERIs:             []*types.ERI{},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            true,
+		},
+		{
+			name: "Existing Secondary RDMA ENI is RDMA and with sufficient queue pairs",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-2"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(8)),
+					Type:                        lo.ToPtr("Secondary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Tags: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetTags{
+						Tag: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetTagsTag{
+							{
+								TagKey:   lo.ToPtr("creator"),
+								TagValue: lo.ToPtr("alibabacloud-erdma-controller"),
+							},
+						},
+					},
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      1,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-2",
+					CardIndex:    0,
+					QueuePair:    8,
+					IsPrimaryENI: false,
+					MAC:          "00:16:3e:00:00:01",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            false,
+		},
+		{
+			name: "Existing Secondary RDMA ENI is RDMA and with sufficient queue pairs no managed",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-2"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(8)),
+					Type:                        lo.ToPtr("Secondary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount:           8,
+			cardCount:                1,
+			expectedERIs:             nil,
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 0,
+			expectedError:            true,
+		},
+		{
+			name: "Existing Secondary RDMA ENI is RDMA and with insufficient queue pairs no managed",
+			existENIs: []*client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSet{
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-1"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("Normal"),
+					QueuePairNumber:             lo.ToPtr(int32(0)),
+					Type:                        lo.ToPtr("Primary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:00"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+				{
+					NetworkInterfaceId:          lo.ToPtr("eni-2"),
+					NetworkInterfaceTrafficMode: lo.ToPtr("HighPerformance"),
+					QueuePairNumber:             lo.ToPtr(int32(4)),
+					Type:                        lo.ToPtr("Secondary"),
+					MacAddress:                  lo.ToPtr("00:16:3e:00:00:01"),
+					Attachment: &client.DescribeNetworkInterfacesResponseBodyNetworkInterfaceSetsNetworkInterfaceSetAttachment{
+						NetworkCardIndex: lo.ToPtr(int32(0)),
+					},
+				},
+			},
+			queuePairCount: 8,
+			cardCount:      1,
+			expectedERIs: []*types.ERI{
+				{
+					ID:           "eni-1",
+					CardIndex:    0,
+					QueuePair:    4,
+					IsPrimaryENI: true,
+					MAC:          "00:16:3e:00:00:00",
+				},
+			},
+			expectedNeedCreate:       []int{},
+			expectedQueuePairPerCard: 4,
+			expectedError:            false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			e := &controller.EriClient{
+				ManagedNonOwned: false,
+			}
+			eris, needCreate, queuePairPerCard, err := e.SelectEriFromExist(tt.existENIs, tt.queuePairCount, tt.cardCount)
+			if tt.expectedError {
+				assert.Error(t, err)
+			} else {
+				if needCreate == nil {
+					needCreate = []int{}
+				}
+				assert.NoError(t, err)
+				assert.Equal(t, tt.expectedERIs, eris)
+				assert.Equal(t, tt.expectedNeedCreate, needCreate)
+				assert.Equal(t, tt.expectedQueuePairPerCard, queuePairPerCard)
+			}
+		})
+	}
+}

